src/System.Net.Primitives/src/System/Net/Cookie.cs

```diff
@@ -0,0 +1,1755 @@
+                }
+                return false;
+            }
+
+            // Check Comment syntax
+            if (Comment != null && !(Comment.Length > 2 && Comment[0] == '\"' && Comment[Comment.Length - 1] == '\"')
+                && (Comment.IndexOfAny(ReservedToValue) != -1))
+            {
+                if (isThrow)
+                {
+                    throw new CookieException(SR.Format(SR.net_cookie_attribute, CommentAttributeName, Comment));
+                }
+                return false;
+            }
+
+            // Check Path syntax
+            if (Path != null && !(Path.Length > 2 && Path[0] == '\"' && Path[Path.Length - 1] == '\"')
+                && (Path.IndexOfAny(ReservedToValue) != -1))
+            {
+                if (isThrow)
+                {
+                    throw new CookieException(SR.Format(SR.net_cookie_attribute, PathAttributeName, Path));
+                }
+                return false;
+            }
+
+            // Check/set domain
+            //
+            // If domain is implicit => assume a) uri is valid, b) just set domain to uri hostname.
+            if (set_default && _domainImplicit == true)
+            {
+                _domain = host;
+            }
+            else
+            {
+                if (!_domainImplicit)
+                {
+                    // Forwarding note: If Uri.Host is of IP address form then the only supported case
+                    // is for IMPLICIT domain property of a cookie.
+                    // The code below (explicit cookie.Domain value) will try to parse Uri.Host IP string
+                    // as a fqdn and reject the cookie.
+
+                    // Aliasing since we might need the KeyValue (but not the original one).
+                    string domain = _domain;
+
+                    // Syntax check for Domain charset plus empty string.
+                    if (!DomainCharsTest(domain))
+                    {
+                        if (isThrow)
+                        {
+                            throw new CookieException(SR.Format(SR.net_cookie_attribute, DomainAttributeName, domain == null ? "<null>" : domain));
+                        }
+                        return false;
+                    }
+
+                    // Domain must start with '.' if set explicitly.
+                    if (domain[0] != '.')
+                    {
+                        if (!(variant == CookieVariant.Rfc2965 || variant == CookieVariant.Plain))
+                        {
+                            if (isThrow)
+                            {
+                                throw new CookieException(SR.Format(SR.net_cookie_attribute, DomainAttributeName, _domain));
+                            }
+                            return false;
+                        }
+                        domain = '.' + domain;
```

Comment:
This will result in an additional box allocation when compiled with the VS2013 csc (because `String.Concat(object, object)` will be used). Not a problem with Roslyn, though, as @stephentoub added an optimization that converts the `char` literal to a `string` literal so that `String.Concat(string, string)` can be used, avoiding the box. I assume corefx will be moving to the Roslyn compiler soon, so might not be worth fixing in the code.